### PR TITLE
Pamatya/appeals 30889

### PR DESCRIPTION
**Pull request by pamatyatake2**
Resolves [APPEALS-30889](https://jira.devops.va.gov/browse/APPEALS-30889)

# Description

1.  Refactor NonCompReportFilter so that it can be reused for non searchable controlled dropdown
2.  Renamed and moved NonCompReportFilter component to RHFControlledDropdown and moved under ReportPage Folder. 
3. Added a two new components -> TimingSpecification.jsx and ReportPageDateSelector.jsx.
4. In TimingSpecification added condition to display ReportPageDateSelector to meet AC.
5. Added a test case to test Timing Specification in ReportPage.test.js

## Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Timing specifications and 'Add Condition' section only appear after ‘Event / Action’ report type is selected in Appeals-30888.
- [ ]  User can select ‘Timing specifications’ range when generating an ‘Event / Action’ report.
       - [ ] Drop-down options for range include:
              - [ ] After
        - [ ] One Date drop-down selector is included to decipher option for ‘After’ option.
              - [ ] Before
         - [ ] One Date drop-down selector is included to decipher option for ‘Before’ option.
              - [ ] Between
         - [ ] Two Date drop-down selectors are included to decipher option for ‘Between’ range.
         - [ ] Last 7 Days
         - [ ] Last 30 Days
         - [ ] Last 365 Days 

## Testing Plan

1. Go to [APPEALS-30889](https://jira.devops.va.gov/browse/APPEALS-30889) or list them below
2. Login as Admin VHA users
3. Navigate to /decision_reviews/vha/report
4. you should see Report Type Dropdown Present with Clear Filter and Generate task report page disabled.
image
6. if you select "Status", you should see both buttons enabled and Conditions Section Appearing. 
image
7. if you select "Event Type/Action" , you will see Timing Specification Section appear. which should have title -> "Timing Specification" and a dropdown with Options listed and behaves like explained in AC. 
**8. Video Showing test:** 
**Updated Video**

https://github.com/department-of-veterans-affairs/caseflow/assets/122557351/0b4f33aa-ccf9-4c5e-b351-66e82ea73e1d

- [ ] For feature branches merging into master: Was this deployed to UAT?

# Frontend
## User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue
image
image
image
image

 BEFORE|AFTER
 ---|---

## Storybook Story
*For Frontend (Presentation) Components*
* [ ] Add a [Storybook](https://github.com/department-of-veterans-affairs/caseflow/wiki/Documenting-React-Components-with-Storybook) file alongside the component file (e.g. create `MyComponent.stories.js` alongside `MyComponent.jsx`)
* [ ] Give it a title that reflects the component's location within the overall Caseflow hierarchy
* [ ] Write a separate story (within the same file) for each discrete variation of the component

# Backend
## Database Changes
*Only for Schema Changes*

* [ ] Add typical timestamps (`created_at`, `updated_at`) for new tables
* [ ] Update column comments; include a "PII" prefix to indicate definite or potential [PII data content](https://github.com/department-of-veterans-affairs/appeals-team/blob/master/caseflow-team/0-how-we-work/pii-handbook.md#what-is-pii)
* [ ] Have your migration classes inherit from `Caseflow::Migration`, especially when adding indexes (use `add_safe_index`) (see [Writing DB migrations](https://github.com/department-of-veterans-affairs/caseflow/wiki/Writing-DB-migrations))
* [ ] Verify that `migrate:rollback` works as desired ([`change` supported functions](https://edgeguides.rubyonrails.org/active_record_migrations.html#using-the-change-method))
* [ ] Perform query profiling (eyeball Rails log, check bullet and fasterer output)
* [ ] For queries using raw sql was an explain plan run by System Team
* [ ] Add appropriate indexes (especially for foreign keys, polymorphic columns, unique constraints, and Rails scopes)
* [ ] Run `make check-fks`; add any missing foreign keys or add to `config/initializers/immigrant.rb` (see [Record associations and Foreign Keys](https://github.com/department-of-veterans-affairs/caseflow/wiki/Record-associations-and-Foreign-Keys))
* [ ] Add `belongs_to` for associations to enable the [schema diagrams](https://department-of-veterans-affairs.github.io/caseflow/task_trees/schema/schema_diagrams) to be automatically updated
* [ ] Document any non-obvious semantics or logic useful for interpreting database data at [Caseflow Data Model and Dictionary](https://github.com/department-of-veterans-affairs/caseflow/wiki/Caseflow-Data-Model-and-Dictionary)

## Integrations: Adding endpoints for external APIs
* [ ] Check that Caseflow's external API code for the endpoint matches the code in the relevant integration repo
  * [ ] Request: Service name, method name, input field names
  * [ ] Response: Check expected data structure
  * [ ] Check that calls are wrapped in MetricService record block
* [ ] Check that all configuration is coming from ENV variables
  * [ ] Listed all new ENV variables in description
  * [ ] Worked with or notified System Team that new ENV variables need to be set
* [ ] Update Fakes
* [ ] For feature branches: Was this tested in Caseflow UAT

# Best practices
## Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [ ] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [ ] No new code climate issues added

### Monitoring
- [ ] Are performance metrics (e.g., response time, throughput) being tracked?
- [ ] Are key application components monitored (e.g., database, cache, queues)?
- [ ] Is there a system in place for setting up alerts based on performance thresholds?

### Logging
- [ ] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?
- [ ] Are logs structured (e.g., using log tags) for easier querying and analysis?
- [ ] Are sensitive data (e.g., passwords, tokens) redacted or omitted from logs?
- [ ] Is log retention and rotation configured correctly?
- [ ] Are logs being forwarded to a centralized logging system if needed?

### Auditing
- [ ] Are user actions being logged for audit purposes?
- [ ] Are changes to critical data being tracked ?
- [ ] Are logs being securely stored and protected from tampering or exposing protected data?

### Error Handling
- [ ] Are errors being caught and handled gracefully?
- [ ] Are appropriate error messages being displayed to users?
- [ ] Are critical errors being reported to an error tracking system (e.g., Sentry, ELK)?
- [ ] Are unhandled exceptions being caught at the application level ?

### Exception Handling
- [ ] Are custom exceptions defined and used where appropriate?
- [ ] Is exception handling consistent throughout the codebase?
- [ ] Are exceptions logged with relevant context and stack trace information?
- [ ] Are exceptions being grouped and categorized for easier analysis and resolution?

### CODE REVIEW

**Review by almorbah — APPROVED** [review]
Looks good to me

**Review by brandondorner — CHANGES_REQUESTED** [review]

File: client/app/nonComp/components/ReportPage/RHFControlledDropdown.jsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+export const RHFControlledDropdownContainer = (props) => {
+  const methods = useFormContext();
+
+  return <RHFControlledDropdown {...methods} props={props} />;
+};
+
+RHFControlledDropdown.propTypes = {
+  control: PropTypes.object,
+  props: PropTypes.object,
```
[the remark this answers is not in the document]
Comment (pamatyatake2) [reply client/app/nonComp/components/ReportPage/RHFControlledDropdown.jsx:39]: I changed the line 34 to as you suggested. But if i remove props validation on line 39, linter is not happy for RHFControllerDropdown, but looks like we don't needs props validation for RHFControlledDropdownContainer

[the remark this answers is not in the document]
Comment (pamatyatake2) [reply client/app/nonComp/components/ReportPage/RHFControlledDropdown.jsx:39]: Also for some reason, when we do {...props} in the RHFControlledDropdownContainer props.header is not valid

File: client/app/nonComp/components/ReportPage/ReportPageDateSelector.jsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+const ReportPageDateSelector = ({ name, label, control }) => {
+  const { field } = useController({
+    control,
+    name,
+    label
+  });
+
+  return (
+    <fieldset style={{ marginTop: '35px' }}>
+      <div key={`${name}_1`}>
```
[the remark this answers is not in the document]
Comment (pamatyatake2) [reply client/app/nonComp/components/ReportPage/ReportPageDateSelector.jsx:17]: done

**Review by brandondorner — APPROVED** [review]
Nice!

**Review by brandondorner — DISMISSED** [review]

**Review by craigrva — APPROVED** [review]

**Pull request merged**